clustered/client/src/test/java/org/ehcache/clustered/client/internal/UnitTestConnectionService.java

```diff
@@ -17,90 +17,384 @@
+      newServer.registerServerEntityService(new CoordinationServerEntityService());
+      newServer.registerClientEntityService(new ClientCoordinationEntityService());
+
+      for (ServerEntityService<?, ?> service : serverEntityServices) {
+        newServer.registerServerEntityService(service);
+      }
+
+      for (EntityClientService<?, ?, ? extends EntityMessage, ? extends EntityResponse> service : clientEntityServices) {
+        newServer.registerClientEntityService(service);
+      }
+
+      if (!this.resources.getResource().isEmpty()) {
+        newServer.registerServiceProvider(new OffHeapResourcesProvider(), new OffHeapResourcesConfiguration(this.resources));
+      }
+
+      for (Map.Entry<ServiceProvider, ServiceProviderConfiguration> entry : serviceProviders.entrySet()) {
+        newServer.registerServiceProvider(entry.getKey(), entry.getValue());
+      }
+
+      return newServer;
+    }
   }
 
-  public static synchronized PassthroughServer server() {
-    return server;
+  public static Collection<Properties> getConnectionProperties(URI uri) {
+    ServerDescriptor serverDescriptor = servers.get(createKey(uri));
+    if (serverDescriptor != null) {
+      return serverDescriptor.getConnections().values();
+    } else {
+      return Collections.emptyList();
+    }
   }
 
   @Override
   public boolean handlesURI(URI uri) {
-    String hostName = uri.getHost();
-    return hostName.endsWith("example.com") || hostName.equals("localhost");
+    return servers.containsKey(createKey(uri));
   }
 
   @Override
   public Connection connect(URI uri, Properties properties) throws ConnectionException {
-    connectionProperties = properties;
-    return server().connectNewClient();
+
+    URI keyURI = createKey(uri);
+    ServerDescriptor serverDescriptor = servers.get(keyURI);
+    if (serverDescriptor == null) {
+      throw new IllegalArgumentException("No server available for " + uri);
+    }
+
+    Connection connection = serverDescriptor.server.connectNewClient();
+    serverDescriptor.add(connection, properties);
+
+    LOGGER.info("Client opened {} to PassthroughServer at {}", formatConnectionId(connection), keyURI);
+
+    /*
+     * Uses a Proxy around Connection so closed connections can be removed from the ServerDescriptor.
+     */
+    return (Connection) Proxy.newProxyInstance(Connection.class.getClassLoader(),
+        new Class[] { Connection.class },
+        new ConnectionInvocationHandler(serverDescriptor, connection));
+  }
+
+  /**
+   * Creates a "key" {@code URI} by dropping the <i>user-info</i>, <i>path</i>, <i>query</i>, and <i>fragment</i>
+   * portions of the {@code URI}.
+   *
+   * @param requestURI the {@code URI} for which the key is to be generated
+   *
+   * @return a {@code URI} instance with the <i>user-info</i>, <i>path</i>, <i>query</i>, and <i>fragment</i> discarded
+   */
+  private static URI createKey(URI requestURI) {
+    try {
```

Comment:
This seems weird... it feels like a detail of the way Ehcache uses URIs is bleeding in to the test framework.  I'd rather see this be done in a test class (statically in code).